lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2262,6 +2262,174 @@ class ArchiveTool : public Tool {
+
+  virtual void configureOutputs(const ConfigureContext& ctx,
+                                const std::vector<Node*>& value) override {}
+
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  StringRef value) override {
+    // No supported attributes.
+    ctx.error("unexpected attribute: '" + name + "'");
+    return false;
+  }
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  ArrayRef<StringRef> values) override {
+    if (name == "expectedOutputs") {
+      expectedOutputs.reserve(values.size());
+      for (auto value : values) {
+        expectedOutputs.emplace_back(value.str());
+      }
+      return true;
+    }
+
+    ctx.error("unexpected attribute: '" + name + "'");
+    return false;
+  }
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  ArrayRef<std::pair<StringRef, StringRef>> values) override {
+    // No supported attributes.
+    ctx.error("unexpected attribute: '" + name + "'");
+    return false;
+  }
+
+  virtual BuildValue getResultForOutput(Node* node,
+                                        const BuildValue& value) override {
+    // If the value was a failed command, propagate the failure.
+    if (value.isFailedCommand() || value.isPropagatedFailureCommand() ||
+        value.isCancelledCommand())
+      return BuildValue::makeFailedInput();
+    if (value.isSkippedCommand())
+      return BuildValue::makeSkippedCommand();
+
+    // Otherwise, this was successful, return the value as-is.
+    return BuildValue::fromData(value.toData());;
+  }
+
+  virtual bool isResultValid(BuildSystem& system,
+                             const BuildValue& value) override {
+    // Always re-run stale file removal.
+    return false;
+  }
+
+  virtual void start(BuildSystemCommandInterface& bsci,
+                     core::Task* task) override {}
+
+  virtual void providePriorValue(BuildSystemCommandInterface&, core::Task*,
+                                 const BuildValue& value) override {
+    hasPriorResult = true;
+    priorValue = BuildValue::fromData(value.toData());
+  }
+
+  virtual void provideValue(BuildSystemCommandInterface&, core::Task*,
+                            uintptr_t inputID,
+                            const BuildValue& value) override {
+    assert(0 && "unexpected API call");
+  }
+
+  virtual BuildValue execute(BuildSystemCommandInterface& bsci,
+                             core::Task* task,
+                             QueueJobContext* context) override {
+    // Nothing to do if we do not have a prior result.
+    if (!hasPriorResult) {
+      bsci.getDelegate().commandStarted(this);
+      bsci.getDelegate().commandFinished(this);
+      return BuildValue::makeStaleFileRemoval(expectedOutputs);
+    }
+
+    if (!priorValue.isStaleFileRemoval()) {
+      return BuildValue::makeFailedCommand();
```

Comment:
I don't think we want to fail the build if the previous command happened to be a failure (like cancelled), do we?